File: template_method/base_post_controller.go
package template_method

import (
	"design_pattern/template_method/custom_errors"
	"errors"
)

type IBasePostController[RQ Requestable, RS Responsible, Dto any] interface {
	MapToDto(requestBody RQ) Dto
	DoService(dto Dto) RS
	CreateResponse(responseBody RS) *Response
}

type BasePostController[RQ Requestable, RS Responsible, Dto any] struct {
	IBasePostController[RQ, RS, Dto]
}

func (b BasePostController[RQ, RS, Dto]) Execute(requestBody RQ) *Response {
	err := b.CheckRequest(requestBody)
	var badRequestError *custom_errors.BadRequestError
	var applicationError *custom_errors.ApplicationError
	if err != nil {
		switch {
		case errors.As(err, &badRequestError):
			return b.createErrorResponse(badRequestError.Error(), 400)
		case errors.As(err, &applicationError):
			return b.createErrorResponse(applicationError.Error(), applicationError.Status)
		default:
			return b.createErrorResponse(err.Error(), 500)
		}
	}

	return b.CreateResponse(b.DoService(b.MapToDto(requestBody)))
}

// CheckRequest Hook Method
func (b BasePostController[RQ, RS, Dto]) CheckRequest(requestBody RQ) error {
	return nil
}

func (b BasePostController[RQ, RS, Dto]) createErrorResponse(message string, status int) *Response {
	return NewResponse(custom_errors.NewErrorResponse(message), status)
}

File: template-method/base-post-controller.ts
import { ApplicationException } from "./exception/application.exception";
import { BadRequestException } from "./exception/bad-request.exception";
import { ErrorResponse } from "./exception/error.response";
import { RequestBody } from "./request-body";
import { Response } from "./response";
import { ResponseBody } from "./response-body";

export abstract class BasePostController<RQ extends RequestBody, RS extends ResponseBody, Dto> {
    execute(requestBody: RQ): Response<any> {
        try {
            this.checkRequest(requestBody);
            return this.createResponse(this.doService(this.mapToDto(requestBody)));
        } catch (error) {
            switch (error.constructor){
                case BadRequestException:
                    return this.createErrorResponse(error.message, 400);
                case ApplicationException:
                    return this.createErrorResponse(error.message, error.status);
                default:
                    return this.createErrorResponse(error.message, 500);
            }
        }
    }

    // Hook Method
    protected checkRequest(requestBody: RQ): void {
    }

    protected abstract mapToDto(requestBody: RQ): Dto;

    protected abstract doService(dto: Dto): RS;

    protected abstract createResponse(responseBody: RS): Response<RS>;

    private createErrorResponse(message: string, status: number): Response<ErrorResponse> {
        return new Response<ErrorResponse>(new ErrorResponse(message), status);
    }
}